doc: Add VSOCK doc page (#4581)
cloud-hypervisor/cloud-hypervisor: docs/vsock.md

# VSOCK support

VSOCK provides a way for guest and host to communicate through a socket. VSOCK sockets support both stream and datagram types.

The `virtio-vsock` is based on the [Firecracker](https://github.com/firecracker-microvm/firecracker/blob/main/docs/vsock.md) implementation, where additional details can be found.

# Basic Terms

## Predefined Values

| CID | Description |
|-----|-------------|
| -1  | Random CID | 
|  0  | Hypervisor | 
|  1  | Loopback | 
|  2  | Host | 

## Socat specific Notes

See the `DATA VALUES` section of `man socat` for the closer info on the data formats. For a quick reference:

| Specifier | Datatype |
|-----------|----------|
|     B     | uint8_t  |
|     S     | uint16_t |
|     I     | uint32_t |

Same man page tells also about the `SOCKET-LISTEN` specificator, where by the last `local-address` component is to contain `the data representation of a sockaddr structure without sa_family`. Consult `linux/vm_sockets.h` on the given system for the details of `struct sockaddr_vm` layout.

> Socat `>=1.7.4` supports `VSOCK-CONNECT` and `VSOCK-LISTEN` address types directly, thus simplifying the usage and eliminating the need to pass a manually crafted `sockaddr` struct data.

Reviewer: Is it worth mentioning about the old way of using `socat` given version >=1.7.4 supports VSOCK directly?

## Kernel Requirements

Host kernel:  CONFIG_VHOST_VSOCK=m

Guest kernel: CONFIG_VIRTIO_VSOCKETS=m 

## Nested VM support

Requirement is Linux `>= 5.5` for the L1 VM.

## Loopback support

Requirement is Linux `>= 5.6`.

> # Establishing Vsock Connection

Reviewer: `## Establishing Vsock Connection`

Vsock device becomes available with `--vsock` option passed by the VM start. As per the previous struct listing, both CID and port are 32-bit values.

Cloud Hypervisor can be invoked for instance as below:

```bash
IMG=jammy-server-cloudimg-amd64.raw
FV=CLOUDHV_EFI.fd
CH=cloud-hypervisor

VSOCK_CID=3
VSOCK_PORT=1234
VSOCK_SOCK=/tmp/ch.vsock

$CH \
	--cpus boot=1 \
	--memory size=4G \
	--kernel $FV \
	--disk path=$IMG \
	--serial tty --console off --net tap= \
	--vsock cid=$VSOCK_CID,socket=$VSOCK_SOCK

> ```

Reviewer: Please make it simpler as this is not a script, just an example of how to run it. And note that `kernel` has been replaced with `firmware` when actually using a firmware.
```
cloud-hypervisor \
	--cpus boot=1 \
	--memory size=4G \
	--firmware CLOUDHV.fd \
	--disk path=jammy-server-cloudimg.raw \
	--vsock cid=3,socket=/tmp/ch.vsock
```

The examples use `socat` to illustrate the VSOCK functionality. However, there are other tools supporting vsock, like [ncat](https://stefano-garzarella.github.io/posts/2019-11-08-kvmforum-2019-vsock/).

> ## Connecting from Host to Guest

Reviewer: `### Connecting from Host to Guest`

The host starts to listen on the defined port:

`$ socat - SOCKET-LISTEN:40:0:S0I1234I3I0`

With newer `socat` versions, this command can be made simpler:

`$ socat - VSOCK-LISTEN:1234`

Once the host is listenint, the guest can send data:

`echo -e "CONNECT $VSOCK_PORT\\nHello from host!" | socat - UNIX-CONNECT:$VSOCK_SOCK`
 

> Note the string `CONNECT <port>` prepended to the actual data. It is possible for the guest to start listening on different ports, thus the specific command is needed to instruct VSOCK to which listener the host wants to connect. It needs to be sent once per a connection. Once the connection established, data transfers can take place directly.

Reviewer: I think it's quite complex to follow when we know socat 1.7.4 supports VSOCK directly. This would simplify greatly the document.

> ## Connecting from Guest to Host

Reviewer: `### Connecting from Guest to Host`

This first requires a listening UNIX socket on the host side. The UNIX socket path has to be constructed by usning the socket path used at the VM launch time with appended `_` and the port number to be used on the guest side. As in the example above, if we'd intended to connect from the guest to the port `1234`, the Unix socket path on the host side will be `/tmp/ch.vsock_1234`. 

Also note, that the CID used on the guest side is the well known CID value `2`.

Listening on the host side:

`$ socat - UNIX-LISTEN:${VSOCK_SOCK}_1234`

From the guest:

`$ echo -e "Hello from guest!" | socat - VSOCK-CONNECT:2:1234`

Or, if we had to pass the sockaddr struct constructed manually:

> `$ echo -e "Hello from guest!" | socat - SOCKET-CONNECT:40:0:S0I1234I2I0`

Reviewer: Ditto about the extra complexity that's not needed with socat 1.7.4 and newer.

> # Links

Reviewer: `## Links`

# VSOCK support

> VSOCKs provide a way for guest and host to communicate through a socket. VSOCK sockets support both stream and datagram types.

Reviewer: ```suggestion
VSOCK provides a way for guest and host to communicate through a socket. VSOCK sockets support both stream and datagram types.
```

# Basic Terms

## C Definitions

The below C definitions illustrate the data structures behind VSOCK and are helpful to understand the calculations used in the following examples.

```C
// bits/socket.h
#define PF_VSOCK        40      /* vSockets.  */
```

```C
// linux/socket.h
typedef unsigned short __kernel_sa_family_t;
```

```C
// linux/vm_sockets.h
struct sockaddr_vm {
        __kernel_sa_family_t svm_family;
        unsigned short svm_reserved1;
        unsigned int svm_port;
        unsigned int svm_cid;
        __u8 svm_flags;
        unsigned char svm_zero[sizeof(struct sockaddr) -
                               sizeof(sa_family_t) -
                               sizeof(unsigned short) -
                               sizeof(unsigned int) -
                               sizeof(unsigned int) -
                               sizeof(__u8)];
};
```

The `sizeof(struct sockaddr)` is 16 bytes, thus the size of the `svm_zero` element would equal 3 in this case.
Note, that the given struct layout needs to be verified on a concrete system. The stucts shown here have been observed

> on a system with glibc 2.35 and kernel 5.17. Be aware, that depending the libc version, the struct layout can differ.

Reviewer: Not sure how relevant this whole section is.

Author: Yep, probably it goes to deep into the glibc level. I've just added the below to the socat notes instead, just in case.

Consult `linux/vm_sockets.h` on the given system for the details of `struct sockaddr_vm` layout.

Thanks

# VSOCK support

VSOCK provides a way for guest and host to communicate through a socket. VSOCK sockets support both stream and datagram types.

The `virtio-vsock` is based on the [Firecracker](https://github.com/firecracker-microvm/firecracker/blob/main/docs/vsock.md) implementation, where additional details can be found.

## What is a CID?

> CID is a 32-bit context identifier describes the source source or destination. In combination with the port, the complete addressing can be achieved to describe multiple listeners running on the same machine.

Reviewer: ```suggestion
CID is a 32-bit context identifier describing the source or destination. In combination with the port, the complete addressing can be achieved to describe multiple listeners running on the same machine.
```

The table below depicts the well known CID values:

| CID | Description |
|-----|-------------|
| -1  | Random CID | 
|  0  | Hypervisor | 
|  1  | Loopback | 
|  2  | Host | 

## Prerequisites

### Kernel Requirements

Host kernel: CONFIG_VHOST_VSOCK

Guest kernel: CONFIG_VIRTIO_VSOCKETS

### Nested VM support

Linux __v5.5__ or newer is required for the L1 VM.

### Loopback support

Linux __v5.6__ or newer is required.

## Establishing Vsock Connection

Vsock device becomes available with `--vsock` option passed by the VM start. Cloud Hypervisor can be invoked for instance as below:

```bash
cloud-hypervisor \
	--cpus boot=1 \
	--memory size=4G \
	--firmware CLOUDHV.fd \
	--disk path=jammy-server-cloudimg.raw \
	--vsock cid=3,socket=/tmp/ch.vsock
```

> The examples use `socat` `>=1.7.4` to illustrate the VSOCK functionality. However, there are other tools supporting vsock, like [ncat](https://stefano-garzarella.github.io/posts/2019-11-08-kvmforum-2019-vsock/).

Reviewer: ```suggestion
The examples use __socat__ `>=1.7.4` to illustrate the VSOCK functionality. However, there are other tools supporting VSOCK, like [ncat](https://stefano-garzarella.github.io/posts/2019-11-08-kvmforum-2019-vsock/).
```

### Connecting from Host to Guest

The host starts to listen on the defined port:

`$ socat - VSOCK-LISTEN:1234`

Once the host is listening, the guest can send data:

`echo -e "CONNECT 1234\\nHello from host!" | socat - UNIX-CONNECT:/tmp/ch.vsock
 

> Note the string `CONNECT <port>` prepended to the actual data. It is possible for the guest to start listening on different ports, thus the specific command is needed to instruct VSOCK to which listener the host wants to connect. It needs to be sent once per a connection. Once the connection established, data transfers can take place directly.

Reviewer: ```suggestion
Note the string `CONNECT <port>` prepended to the actual data. It is possible for the guest to start listening on different ports, thus the specific command is needed to instruct VSOCK to which listener the host wants to connect. It needs to be sent once per connection. Once the connection established, data transfers can take place directly.
```

### Connecting from Guest to Host

> This first requires a listening UNIX socket on the host side. The UNIX socket path has to be constructed by usning the socket path used at the VM launch time with appended `_` and the port number to be used on the guest side. As in the example above, if we'd intended to connect from the guest to the port `1234`, the Unix socket path on the host side will be `/tmp/ch.vsock_1234`. 

Reviewer: ```suggestion
This first requires a listening UNIX socket on the host side. The UNIX socket path has to be constructed by using the socket path used at the VM launch time with appended `_` and the port number to be used on the guest side. As in the example above, if we'd intended to connect from the guest to the port `1234`, the Unix socket path on the host side would be `/tmp/ch.vsock_1234`.
```

> Also note, that the CID used on the guest side is the well known CID value `2`.

Reviewer: ```suggestion
Also note that the CID used on the guest side is the well known CID value `2`.
```
